### PR TITLE
 [IMP] mail, sms, snailmail, test_mail(_full): adapt mail.notification

### PR DESCRIPTION
**Pull request by seb-odoo**
Purpose is to have more common code for failure and notifications, with less
override in `sms` and `snailmail`.

task-2176017

Enterprise https://github.com/odoo/enterprise/pull/9140
Upgrade https://github.com/odoo/upgrade/pull/918

### CODE REVIEW

File: addons/snailmail/static/src/js/thread_widget.js [hunk 1]
```diff
@@ -5,39 +5,18 @@ var ThreadWidget = require('mail.widget.Thread');
 var Dialog = require('web.Dialog');
 
 var core = require('web.core');
-var QWeb = core.qweb;
 var _t = core._t;
 
 ThreadWidget.include({
     dependencies: ['bus_service'],
```
Comment (alexkuhn) [inline addons/snailmail/static/src/js/thread_widget.js]: could also remove this, this is useless

File: addons/snailmail/static/src/js/thread_widget.js [hunk 2]
```diff
@@ -57,17 +36,25 @@ ThreadWidget.include({
...
-            buttons: [
-                {text: _t("Re-send letter"), classes: 'btn-primary', close: true, click: _.bind(message.resendLetter, message)},
-                {text: _t("Cancel letter"), close: true, click: _.bind(message.cancelLetter, message)},
-                {text: _t("Close"), close: true},
-            ],
+            buttons: [{
+                text: _t("Re-send letter"),
+                classes: 'btn-primary',
+                close: true,
+                click: _.bind(message.resendLetter, message),
```
Comment (alexkuhn) [inline addons/snailmail/static/src/js/thread_widget.js]: prefer `(...args) => f(...args)` over `f.bind(r)` or `_.bind(f)`

File: addons/snailmail/static/src/xml/thread.xml [hunk 3]
```diff
@@ -1,37 +1,49 @@
...
+    <t t-inherit="mail.widget.Thread.Message.MailTooltip" t-inherit-mode="extension">
+        <!--
+            It was decided that the information displayed for snaimail messages has to be
+            different than for standard messages, see task-1907998.
+        -->
+        <xpath expr="//span[@name='notification_status']" position="replace">
+            <t t-if="notification.notification_type === 'snailmail'">
+                <span class="d-inline-block text-center o_thread_tooltip_icon">
+                    <t t-if="notification.notification_status === 'sent'">
+                        <i class='fa fa-check o_thread_tooltip_snailmail_icon' title="Sent" role="img" aria-label="Sent"/>
```
Comment (alexkuhn) [inline addons/snailmail/static/src/xml/thread.xml]: `"` instead of `'` around attribute value in XML

File: addons/snailmail/static/tests/chatter_snailmail_tests.js [hunk 4]
```diff
@@ -106,10 +102,12 @@ QUnit.module('Chatter', {
...
 
 QUnit.test('Sent', async function (assert) {
-    assert.expect(7);
+    assert.expect(6);
 
-    this.data['mail.message'].records[0].snailmail_status = 'sent';
-    this.data['mail.message'].records[0].snailmail_error = false;
+    this.data['mail.message'].records[0].notifications = [{
+        'notification_type': 'snailmail',
+        'notification_status': 'sent',
```
Comment (alexkuhn) [inline addons/snailmail/static/tests/chatter_snailmail_tests.js]: no quotes around prop name when not necessary

File: addons/test_mail/tests/test_performance.py [hunk 5]
```diff
@@ -697,6 +697,116 @@ def test_complex_tracking_template(self):
...
+        """Test performance of `_message_read_dict_postprocess` and of
+        `message_format` with multiple messages with multiple attachments,
+        different authors, various notifications, and different tracking values.
+        Those messages might not make sense functionally but they are crafted to
+        cover as much of the code as possible in regard to number of queries.
+        """
+        name_field = self.env['ir.model.fields']._get(self.umbrella._name, 'name')
+        customer_id_field = self.env['ir.model.fields']._get(self.umbrella._name, 'customer_id')
+
+        messages = self.env['mail.message'].sudo().with_context(pudb=True).create([{
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/test_mail/tests/test_performance.py]: Good catch, remaining of me debugging

File: addons/mail/models/mail_notification.py [hunk 6]
```diff
@@ -82,3 +82,32 @@ def _gc_notifications(self, max_age_days=180):
...
+    def _status_for_message_list(self):
+        """Returns the status of notifications allowed on the message list."""
+        return ['bounce', 'exception', 'canceled']
+
+    def _filtered_for_message_list(self):
+        """Returns only the notifications to show on the message list."""
+        return self.filtered(
+            lambda n:
+                n.notification_type in self ._types_for_message_list() and
+                (n.notification_status in self ._status_for_message_list() or n.res_partner_id.partner_share)
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/models/mail_notification.py]: typo

File: addons/mail/static/src/js/models/messages/abstract_message.js [hunk 7]
```diff
@@ -163,6 +163,35 @@ var AbstractMessage =  Class.extend({
...
+     * Gets the class to use as the notification icon.
+     *
+     * @returns {string}
+     */
+    getNotificationIcon: function () {
+        let fatype = '';
+        if (!this.hasNotificationsError()) {
+            fatype = '-o';
+        }
+        return 'fa fa-envelope' + fatype;
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/js/models/messages/abstract_message.js]: Good idea

File: addons/mail/static/src/js/models/messages/abstract_message.js [hunk 8]
```diff
@@ -163,6 +163,35 @@ var AbstractMessage =  Class.extend({
...
+    },
+    /**
+     * Gets the text to display next to the notification icon.
+     *
+     * @returns {string}
+     */
+    getNotificationText: function () {
+        return '';
+    },
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/js/models/messages/abstract_message.js]: done

File: addons/mail/static/src/js/models/messages/abstract_message.js [hunk 9]
```diff
@@ -163,6 +163,35 @@ var AbstractMessage =  Class.extend({
     getNonImageAttachments: function () {
         return _.difference(this.getAttachments(), this.getImageAttachments());
     },
+    /**
+     * Gets the notifications of this message.
+     * By default messages do not have notifications.
+     *
+     * @returns {Object[]}
+     */
+    getNotifications: function () {
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/js/models/messages/abstract_message.js]: done

File: addons/mail/static/src/js/models/messages/abstract_message.js [hunk 10]
```diff
@@ -236,6 +256,28 @@ var AbstractMessage =  Class.extend({
...
+     * @returns {boolean}
+     */
+    hasNotificationsError: function () {
+        if (_.some(this.getNotifications(), notif =>
+            notif.notification_status === 'exception' ||
+            notif.notification_status === 'bounce'
+        )) {
+            return true;
+        }
+        return false;
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/js/models/messages/abstract_message.js]: done (with return directly not intermediary variable)

File: addons/mail/static/src/js/models/messages/message.js [hunk 11]
```diff
@@ -24,8 +24,7 @@ var Message =  AbstractMessage.extend(Mixins.EventDispatcherMixin, ServicesMixin
      * @param {Object} data
      * @param {string} [data.body = ""]
      * @param {(string|integer)[]} [data.channel_ids]
-     * @param {Object[]} [data.customer_email_data]
-     * @param {string} [data.customer_email_status]
+     * @param {Object[]} [data.notifications]
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/js/models/messages/message.js]: done

File: addons/mail/static/src/js/models/messages/message.js [hunk 12]
```diff
@@ -202,6 +171,12 @@ var Message =  AbstractMessage.extend(Mixins.EventDispatcherMixin, ServicesMixin
         }
         return this._emailFrom;
     },
+    /**
+     * @override
+     */
+    getNotifications: function () {
+        return Object.values(this._notifications);
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/js/models/messages/message.js]: ok, added comment

File: addons/mail/static/src/xml/thread.xml [hunk 13]
```diff
@@ -452,18 +450,17 @@
...
     -->
     <t t-name="mail.widget.Thread.Message.MailTooltip">
-        <div t-foreach="data" t-as="customerEmailData">
-            <span class="d-inline-block text-center o_thread_tooltip_icon">
-                <i t-if="customerEmailData[2] === 'sent'" class='fa fa-check' title="Sent" role="img" aria-label="Sent"/>
-                <i t-if="customerEmailData[2] === 'bounce'" class='fa fa-exclamation text-danger' title="Bounced" role="img" aria-label="Bounced"/>
-                <i t-if="customerEmailData[2] === 'exception'" class='fa fa-exclamation text-danger' title="Error" role="img" aria-label="Error"/>
-                <i t-if="customerEmailData[2] === 'ready'" class='fa fa-send-o' title="Ready" role="img" aria-label="Ready"/>
-                <i t-if="customerEmailData[2] === 'canceled'" class='fa fa-trash-o' title="Canceled" role="img" aria-label="Canceled"/>
+        <div t-foreach="notifications" t-as="notification">
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/xml/thread.xml]: done

File: addons/mail/models/mail_message.py [hunk 14]
```diff
@@ -1162,47 +1110,36 @@ def _get_message_format_fields(self):
...
+        """
+        return {
+            message.id: {
+                'message_id': message.id,
+                'model_name': message.env['ir.model']._get(message.model).display_name,
+                'res_id': message.res_id,
+                'model': message.model,
+                'last_message_date': message.date,
+                'message_type': message.message_type,
+                'notifications': message.notification_ids._filtered_for_message_list()._format_notifications(),
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/models/mail_message.py]: Agree, the goal is to ease that migration to OWL not do the complete refactoring without it.

File: addons/mail/static/src/xml/thread.xml [hunk 15]
```diff
@@ -324,13 +324,11 @@
...
                     </t>
-                    <span t-if="options.displayEmailIcons and (message.hasCustomerEmailData() or ['exception', 'bounce'].indexOf(message._customerEmailStatus) > -1)" class="o_thread_tooltip_container">
-
-                        <t t-set="fatype" t-value="''"/>
-                        <t t-if="message.getCustomerEmailStatus() === 'sent' or message.getCustomerEmailStatus() === 'ready'">
-                            <t t-set="fatype" t-value="'-o'"/>
-                        </t>
-                        <i t-att-class="'o_thread_tooltip o_thread_message_email o_thread_message_email_' + message.getCustomerEmailStatus() + ' fa fa-envelope' + fatype" t-att-data-message-id="message.getID()"/>
+                    <span t-if="options.displayNotificationIcons and message.hasNotifications()" class="o_thread_tooltip_container">
+                        <span name="notification_icon" t-att-class="'d-inline-flex align-items-center o_thread_tooltip o_thread_message_notification ' + (message.hasNotificationsError() ? 'o_thread_message_notification_error' : '')" t-att-data-message-id="message.getID()" t-att-data-message-type="message.getType()">
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/xml/thread.xml]: ok

File: addons/mail/tests/common.py [hunk 16]
```diff
@@ -494,6 +484,16 @@ def assertMailNotifications(self, messages, recipients_info):
 
         return done_msgs, done_notifs
 
+    def assertMailBusNotifications(self, message):
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/tests/common.py]: Part of the other PR but ok to add docstring. We'll see if TDE has any comment about the name.

File: addons/sms/static/src/js/mail_failure.js [hunk 17]
```diff
@@ -16,14 +16,36 @@ MailFailure.include({
      */
     getPreview: function () {
         var preview = this._super.apply(this, arguments);
-        if (this._failureType === 'sms') {
+        if (this._messageType === 'sms') {
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/sms/static/src/js/mail_failure.js]: done

File: addons/sms/static/src/js/message.js [hunk 18]
```diff
@@ -1,83 +1,33 @@
 odoo.define('sms.model.Message', function (require) {
 "use strict";
 
+var core = require('web.core');
+var _t = core._t;
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/sms/static/src/js/message.js]: done

File: addons/snailmail/static/src/js/thread_widget.js (hunk 1 above)
Comment (seb-odoo) [reply addons/snailmail/static/src/js/thread_widget.js]: ok

File: addons/snailmail/static/src/js/thread_widget.js (hunk 2 above)
Comment (seb-odoo) [reply addons/snailmail/static/src/js/thread_widget.js]: ok

File: addons/snailmail/static/src/xml/thread.xml (hunk 3 above)
Comment (seb-odoo) [reply addons/snailmail/static/src/xml/thread.xml]: ok

File: addons/snailmail/static/tests/chatter_snailmail_tests.js (hunk 4 above)
Comment (seb-odoo) [reply addons/snailmail/static/tests/chatter_snailmail_tests.js]: ok

Comment (seb-odoo) [reply addons/snailmail/static/tests/chatter_snailmail_tests.js]: Actually not ok, kept as string since they are data that are "coming" from the server so they have to keep the snake case, but we would like to lint camel case some day ideally, which would fail in this case if it wasn't as string.

File: addons/mail/static/src/xml/thread.xml [hunk 19]
```diff
@@ -452,18 +450,17 @@
...
-                <i t-if="customerEmailData[2] === 'exception'" class='fa fa-exclamation text-danger' title="Error" role="img" aria-label="Error"/>
-                <i t-if="customerEmailData[2] === 'ready'" class='fa fa-send-o' title="Ready" role="img" aria-label="Ready"/>
-                <i t-if="customerEmailData[2] === 'canceled'" class='fa fa-trash-o' title="Canceled" role="img" aria-label="Canceled"/>
+        <div t-foreach="notifications" t-as="notification">
+            <span name="notification_status" class="d-inline-block text-center o_thread_tooltip_icon">
+                <i t-if="notification.notification_status === 'sent'" class='fa fa-check' title="Sent" role="img" aria-label="Sent"/>
+                <i t-if="notification.notification_status === 'bounce'" class='fa fa-exclamation text-danger' title="Bounced" role="img" aria-label="Bounced"/>
+                <i t-if="notification.notification_status === 'exception'" class='fa fa-exclamation text-danger' title="Error" role="img" aria-label="Error"/>
+                <i t-if="notification.notification_status === 'ready'" class='fa fa-clock-o' title="Ready" role="img" aria-label="Ready"/>
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/xml/thread.xml]: I'll revert the icon, I don't want too much functional change in this PR.

**Comment by seb-odoo** [comment]
@alexkuhn could you check this again? I've made the changes and rebased.

File: addons/snailmail/models/mail_notification.py [hunk 20]
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
+
+from odoo import api, fields, models
+
+
+class Notification(models.Model):
+    _inherit = 'mail.notification'
+
+    notification_type = fields.Selection(selection_add=[('snailmail', 'Snailmail')])
```
Comment (tde-banana-odoo) [inline addons/snailmail/models/mail_notification.py]: We have a task to reduce a bit notification footprint, and it includes selection keys. Some characters, for millions of notifications, make quite an heavy db. I therefore propose to reduce to 'snail' as it is normally specific enough to avoid name conflicts while staying short.

File: addons/test_mail_full/tests/test_sms_performance.py [hunk 21]
```diff
@@ -51,7 +51,7 @@ def setUp(self):
     def test_message_sms_record_1_partner(self):
         record = self.test_record.with_user(self.env.user)
         pids = self.customer.ids
-        with self.mockSMSGateway(), self.assertQueryCount(employee=21):  # test_mail_enterprise: 21
+        with self.mockSMSGateway(), self.assertQueryCount(employee=26):  # test_mail_enterprise: 26
```
Comment (tde-banana-odoo) [inline addons/test_mail_full/tests/test_sms_performance.py]: Where does those additional queries come from ?

File: addons/mail/models/mail_notification.py [hunk 22]
```diff
@@ -82,3 +82,34 @@ def _gc_notifications(self, max_age_days=180):
             ('notification_status', 'in', ('sent', 'canceled'))
         ]
         return self.search(domain).unlink()
+
+    @api.model
+    def _types_for_web_client(self):
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_notification.py]: I don't really like the naming and _status_for_web_client does not seem to be used. I think we should not support exotic statuses for failed notification: bounce, exception and canceled should be sufficient (allowing to remove one method).

Concerning _types_for_web_client: shouldn't we exclude inbox instead of adding them one by one for each new selection key ? Maybe it lessens ease of override for external people but if it works for our standard use cases ... less methods, less code, less tests :D .

File: addons/sms/models/sms_sms.py [hunk 23]
```diff
@@ -135,6 +135,7 @@ def _postprocess_iap_sent_sms(self, iap_results, failure_reason=None, delete_all
                         'failure_type': self.IAP_TO_SMS_STATE[state] if state != 'success' else False,
                         'failure_reason': failure_reason if failure_reason else False,
                     })
+        self.mail_message_id._notify_message_notification_update()
```
Comment (tde-banana-odoo) [inline addons/sms/models/sms_sms.py]: Why this change ?

File: addons/test_mail_full/tests/test_sms_performance.py (hunk 21 above)
Comment (seb-odoo) [reply addons/test_mail_full/tests/test_sms_performance.py]: This is related to your question in `_postprocess_iap_sent_sms`. There was a missing notification that was added so it has to fetch data for it.

File: addons/sms/models/sms_sms.py (hunk 23 above)
Comment (seb-odoo) [reply addons/sms/models/sms_sms.py]: The web client was out of date in that case because the notification wasn't sent after an update of the notification record just above.

File: addons/snailmail/models/mail_notification.py [hunk 24]
```diff
@@ -0,0 +1,22 @@
...
+    letter_id = fields.Many2one('snailmail.letter', string='Snailmail Letter', index=True, ondelete='set null')
+    failure_type = fields.Selection(selection_add=[
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/snailmail/models/mail_notification.py]: Ok for me, I'll change to your suggestions.

File: addons/mail/models/mail_notification.py (hunk 22 above)
Comment (seb-odoo) [reply addons/mail/models/mail_notification.py]: Ok to remove `_status_for_web_client`, I originally needed it for snaimail but then I decided to use "exception" as you suggest.

`_types_for_web_client` can indeed be replaced by just excluding "inbox" in the main method, I asked myself the same question. I went for the separate method returning a specific list to give more rooms to customization and to make it less prone to side effects, but I don't mind to change if you prefer less code here.

File: addons/snailmail/models/mail_notification.py (hunk 20 above)
Comment (seb-odoo) [reply addons/snailmail/models/mail_notification.py]: Agree, `snail` is actually much easier to find in the code than `snailmail` due to it being a unique word.

File: addons/snailmail/models/mail_notification.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/snailmail/models/mail_notification.py]: Just "fmt" seems very unclear to me all I prefer to shorten it to "format" (which is acceptable since it's a similar length to others).

File: addons/test_mail_full/tests/test_sms_performance.py (hunk 21 above)
Comment (seb-odoo) [reply addons/test_mail_full/tests/test_sms_performance.py]: 1. read message
2. read ir_model based on model name provided by message (because of `model_name` in notif format)
3. search partner associated to notification (partner name for notif format)
4. insert bus

**Comment by seb-odoo** [comment]
@tde-banana-odoo Thank you for the first quick check. All changes have been applied, let me know.

File: addons/snailmail/static/src/js/thread_widget.js [hunk 25]
```diff
@@ -77,46 +63,17 @@ ThreadWidget.include({
...
             $content: $('<div>').html(content),
-            buttons: [
-                {text: _t("Cancel letter"), classes: 'btn-primary', close: true, click: _.bind(message.cancelLetter, message)},
-                {text: _t("Close"), close: true},
-            ],
+            buttons: [{
+                text: _t("Cancel letter"),
+                classes: 'btn-primary',
+                close: true,
+                click: _.bind(message.cancelLetter, message),
```
Comment (alexkuhn) [inline addons/snailmail/static/src/js/thread_widget.js]: (...args) => message.cancel(...args)

File: addons/mail/static/src/js/models/messages/abstract_message.js [hunk 26]
```diff
@@ -236,6 +255,25 @@ var AbstractMessage =  Class.extend({
...
+    hasNotifications() {
+        return this.getNotifications().length > 0;
+    },
+    /**
+     * States whether this message has notifications that are in error.
+     *
+     * @returns {boolean}
+     */
+    hasNotificationsError() {
+        return _.some(this.getNotifications(), notif =>
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/js/models/messages/abstract_message.js]: ok

File: addons/mail/static/src/js/services/mail_notification_manager.js [hunk 27]
```diff
@@ -299,7 +299,7 @@ MailManager.include({
      * @param {string} datas.elements[].notifications[0] sending state of a mail
      *   failure (e.g. 'exception').
      */
-    _handlePartnerMailFailureNotification: function (datas) {
+    _handlePartnerMessageNotificationUpdate(datas) {
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/js/services/mail_notification_manager.js]: ok

File: addons/mail/static/src/js/thread_widget.js [hunk 28]
```diff
@@ -583,7 +583,7 @@ var ThreadWidget = Widget.extend({
      * @private
      * @param {MouseEvent} ev
      */
-    _onClickEmailException: function (ev) {
+    _onClickMessageNotificationError: function (ev) {
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/js/thread_widget.js]: ok

File: addons/sms/static/src/js/thread_widget.js [hunk 29]
```diff
@@ -2,88 +2,26 @@ odoo.define('sms.widget.Thread', function (require) {
...
-    _onClickSMSError: function (ev) {
-        var messageID = $(ev.currentTarget).data('message-id');
-        this.do_action('sms.sms_resend_action', {
-            additional_context: {
-                default_mail_message_id: messageID
-            }
-        });
+    _onClickMessageNotificationError(ev) {
+        if ($(ev.currentTarget).data('message-type') === 'sms') {
+            var messageID = $(ev.currentTarget).data('message-id');
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/sms/static/src/js/thread_widget.js]: ok

File: addons/snailmail/static/src/js/thread_widget.js (hunk 25 above)
Comment (seb-odoo) [reply addons/snailmail/static/src/js/thread_widget.js]: ok

**Comment by seb-odoo** [comment]
All changes applied.

**Comment by seb-odoo** [comment]
I have rebased and wrote proper migration scripts/tests. The plan is to merge as soon as they are approved (by runbot and by the migration team).

**Comment by tde-banana-odoo** [comment]
Why do you add queries ?

**Comment by seb-odoo** [comment]
Quoting my answer to the same question about query:
> seb-odoo on Mar 10 Author Member
> 
> This is related to your question in _postprocess_iap_sent_sms. There was a missing notification that was added so it has to fetch data for it.

> seb-odoo on Mar 10 Author Member
> 
>     read message
>     read ir_model based on model name provided by message (because of model_name in notif format)
>     search partner associated to notification (partner name for notif format)
>     insert bus
> 
>

**Comment by tde-banana-odoo** [comment]
And we think it is necessary to add so much queries for that ?

**Comment by tde-banana-odoo** [comment]
Apart from that you basically fucked up my current branch but that's not your fault :D .

**Comment by seb-odoo** [comment]
> And we think it is necessary to add so much queries for that ?

Is the missing notification necessary? For me absolutely yes, as not having it leads to the user doing an action and not having any clue if anything happened or not.

Are all of those queries necessary for that notification? Well I just listed each of them, so if we want to reduce them, we need to send less information with the notification in general (maybe, but out of scope in my opinion), or take more advantage of the cache (only if possible, and also out of scope).
